Fix #164 - Enable run/debug on local server

briandealwis opened the pull request:
This feels a bit hacky:
  - explicitly specify timezone=UTC to avoid for the `NoSuchMethodException` on Java8
  - copy over module publishing from GPE; run `gcloud` from the published locations
  - run debug launch with `quiet=y` to avoid spurious messages on stdout, which interferes with `dev_appserver`'s communication with the running web server
  - add support for launching multiple `war`s (App Engine services): have to have same App Engine Application/Project ID, and distinct `<service>` names

PTAL @elharo — I might have a chance to look at this further tonight.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+
+  @Override
+  public int getOrder() {
+    // Hmm, shouldn't the publish tasks come later?
```
todo

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+   */
+  public CloudSdkPublishOperation(CloudSdkServerBehaviour server, int kind, IModule[] modules,
+      int deltaKind) {
+    super("Publish to server", "Publish a modules to App Engine Development Server");
```
"a modules" --> "a module"

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+
+  @Override
+  public void execute(IProgressMonitor monitor, IAdaptable info) throws CoreException {
+    // TODO(amitin): use more advanced key to store modules publish locations? Because a dependent
```
who is amitin?

briandealwis replied:
Dunno; I'll remove the name.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+  public void execute(IProgressMonitor monitor, IAdaptable info) throws CoreException {
+    // TODO(amitin): use more advanced key to store modules publish locations? Because a dependent
+    // java project (added as child modules and published as jar) cannot present in more than one
+    // parent modules.
```
modules --> module

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+      // root modules
+      publishDir(deployPath, statusList, monitor);
+    } else {
+      // BSD: NEED TO UNDERSTAND SITUATION FOR REMAINING CODE
```
?

briandealwis replied:
For each module-set added to a server, we create a corresponding `CloudSdkPublishOperation` (done by `CloudSdkPublishTaskDelegate`).  I'm not clear when and why we might get multiple modules in a set. Perhaps .wars including other war projects?

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+        IPath deployPath = server.getModuleDeployDirectory(modules[i]);
+        IWebModule webModule = (IWebModule) modules[i].loadAdapter(IWebModule.class, monitor);
+        if (webModule == null) {
+          statusList.add(newErrorStatus("Not a Web modules: " + modules[i].getName()));
```
modules --> module

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+          // no uri is OK for removed modules
+          if (deltaKind != ServerBehaviourDelegate.REMOVED) {
+            statusList
+                .add(newErrorStatus("Cannot get URI for modules: " + modules[i + 1].getName()));
```
modules --> module

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+          }
+        }
+      }
+      // BSD: Disabled saving and loading module locations
```
?

briandealwis replied:
The original code in GPE wrote out these modules to a separate file; I haven't copied those over.  I wonder why they didn't use server attributes instead.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+      } else {
+          publishDir(deployPath, moduleUrls, statusList, monitor, childModule);
+      }
+      // FIXME: server.saveModulePublishLocations(moduleUrls);
```
still relevant?

briandealwis replied:
This is one of the elements that I commented out

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+      IProgressMonitor monitor, IModule childModule) {
+    boolean isMoving = false;
+    // check older publish
+    String oldURI = (String) mapping.get(childModule.getId());
```
oldUri

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+    boolean isMoving = false;
+    // check older publish
+    String oldURI = (String) mapping.get(childModule.getId());
+    String jarURI = path.toOSString();
```
jarUri

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+  /**
+   * Publish modules as directory.
+   */
+  private void publishDir(IPath path, List<IStatus> statusList, IProgressMonitor monitor)
```
publishDirectory

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,330 @@
+      IProgressMonitor monitor, IModule childModule) throws CoreException {
+    boolean isMoving = false;
+    // check older publish
+    String oldURI = (String) mapping.get(childModule.getId());
```
oldUri, dirUri

elharo commented:
some tests are failing

briandealwis commented:
PTAL @elharo.
  - I moved the Java8 timezone hack into the launch delegate so that the tests pass (and avoid putting in workarounds into tests).
  - I discovered that we can only debug a single module at a time, but we can launch multiple modules in non-debug mode.
  - I added support to the LaunchDelegate, which actually performs the launch, to add any flags specified on the Launch Configuration dialog's Server Flags tab.  Unfortunately the launch configuration's defaults aren't set until the user actually opens the Launch Configuration dialog as it might otherwise be a great place to put the timezone workaround flag.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java:
```diff
@@ -112,11 +114,19 @@ public void serverChanged(ServerEvent event) {
+
       Activator.logInfo("Executing: " + commands);
 
       Process p = Runtime.getRuntime().exec(commands, null);
```
p --> process

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,329 @@
+    }
+    IStatus[] children = statusList.toArray(new IStatus[statusList.size()]);
+    throw new CoreException(
+        new MultiStatus(PLUGIN_ID, 0, children,
```
could the new MultiStatus part fit on one line?

briandealwis replied:
I tried but the GoogleStyle settings I have loaded don't seem to like it.  We'll get this fixed with #109.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,329 @@
+  @Override
+  public int getOrder() {
+    // TODO: shouldn't publishing tasks come later?
+    // may be necessary to work with GWT compilation
```
but we aren't doing GWT compilation

briandealwis replied:
I mentioned that Brandon had been experimenting with rewriting GWT as a WTP facet that performs the GWT compilation as part of the war creation process.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,329 @@
+
+  @Override
+  public void execute(IProgressMonitor monitor, IAdaptable info) throws CoreException {
+    // TODO: use more advanced key to store modules publish locations? Because a dependent
```
I'm not sure what to do with these legacy todos. On the one hand, I don't want to just lose the info. OTOH, I doubt they're still relevant or useful.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java:
```diff
@@ -0,0 +1,329 @@
+  }
+
+  /**
+   * Publish modules by zipping it into JAR file.
```
zipping them

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -192,4 +203,43 @@ private void stopDevAppServer() {
+  }
+
+  /**
+   * Convenient method allowing access to protected method in superclass.
```
Convenience

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -192,4 +203,43 @@ private void stopDevAppServer() {
+  }
+
+  /**
+   * Convenient method allowing access to protected method in superclass.
```
Convenience

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -192,4 +203,43 @@ private void stopDevAppServer() {
+  }
+
+  /**
+   * Convenient accessor to protected member in superclass.
```
Convenience

elharo commented:
a few minor nits, but otherwise LGTM
